hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,307 @@
+use crate::indexed_merkle_proof::IndexedMerkleProof;
+
+use casper_types::{
+    allocate_buffer,
+    bytesrepr::{Bytes, FromBytes, ToBytes},
+};
+
+use crate::{error, Digest};
+
+#[derive(PartialEq, Debug, schemars::JsonSchema, serde::Serialize, serde::Deserialize)]
+#[schemars(with = "String", description = "Hex-encoded hash digest.")]
+#[serde(deny_unknown_fields)]
+pub struct ChunkWithProof {
+    proof: IndexedMerkleProof,
+    chunk: Vec<u8>,
+}
+
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, casper_types::bytesrepr::Error> {
+        let mut result = allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut Bytes::from(self.chunk.as_slice()).to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), casper_types::bytesrepr::Error> {
+        let ((proof, chunk), remainder) = <(IndexedMerkleProof, Bytes)>::from_bytes(bytes)?;
+        Ok((
+            ChunkWithProof {
+                proof,
+                chunk: chunk.into(),
+            },
+            remainder,
+        ))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    pub const CHUNK_SIZE_BYTES: usize = 10;
+
+    #[cfg(not(test))]
+    pub const CHUNK_SIZE_BYTES: usize = 1_048_576; // 2^20
+
+    #[allow(unused)]
+    fn new(data: &[u8], index: u64) -> Result<Self, error::MerkleConstructionError> {
+        if data.len() < Self::CHUNK_SIZE_BYTES * (index as usize) {
```

Comment:
Hmm, this is something that I'd like to get an opinion from @xcthulhu before making changes, since this code has been originally introduced in the following commit.

https://github.com/xcthulhu/casper-node/commit/cbd173fa9121833560c1546dcdf7c1ed15b92bdd#diff-9bcc1d978366a325e4a868daa50342befde8fbf7197b9f3e5b7046b47d8f9a2fR15

Reply:
I'm thinking that if e.g. `CHUNK_SIZE_BYTES` is 10, and `data.len()` is 20, then the data will be split into two chunks of length 10, numbers 0 and 1. Chunk number 2 would be invalid, even though `data.len() == Self::CHUNK_SIZE_BYTES * 2`.

But my impression is that this check is duplicated and `IndexedMerkleProof::new` would already return `IndexOutOfBounds` anyway. Because if none of the chunks matches the index we'll end up in the `Some(Hash(_))` case there.

Reply:
> I'm thinking that if e.g. CHUNK_SIZE_BYTES is 10, and data.len() is 20, then the data will be split into two chunks of length 10, numbers 0 and 1. Chunk number 2 would be invalid, even though data.len() == Self::CHUNK_SIZE_BYTES * 2.

Yes, sounds right. `20 < 20` still yields `false` while it shouldn't. However, changing to `<=` causes some tests to fail.

>But my impression is that this check is duplicated

Yes, I too think it's duplicated. However, the following call:
```
ChunkWithProof::new(&[], 1)
```

will eventually produce an `IndexedMerkleProof` with count `1`, not `0`, due to injecting a digest:
```
if data.is_empty() {
    (IndexedMerkleProof::new([Digest::hash(&[])], index)?, vec![])
}
```

Hence, the `IndexedMerkleProof::new()` will indeed return `MerkleConstructionError::IndexOutOfBounds`, but instead of count `0`, it'll incorrectly report `1`. This causes another test to fail.

<br />
<br />

I need to give it a second thought of what is actually expected from the `ChunkWithProof::new()` given the different inputs, (including empty one) and eventually align the tests.

Reply:
But why move that check one level up at all?

Before, `new` did the index check itself, returned an appropriate error if the index was invalid, and never crashed.

With https://github.com/casper-network/casper-node/commit/8ab10efe991a45fe3f1168799f1bef5ce7f8e1cc, `new` would now _crash_ on an invalid index, and instead the caller does the check. That seems less safe to me.

Why not just keep the check in `new`, and remove the redundant check in the caller?

> However, changing to `<=` causes some tests to fail.

That was just because of the "empty" case, was it?